zkvm/src/signature/musig.rs

```diff
@@ -0,0 +1,176 @@
+#![allow(non_snake_case)]
+
+use crate::errors::VMError;
+use crate::signature::VerificationKey;
+use crate::transcript::TranscriptProtocol;
+use curve25519_dalek::constants::RISTRETTO_BASEPOINT_POINT;
+use curve25519_dalek::ristretto::RistrettoPoint;
+use curve25519_dalek::scalar::Scalar;
+use merlin::Transcript;
+
+#[derive(Clone)]
+pub struct PubKey(pub RistrettoPoint);
+
+#[derive(Clone)]
+pub struct PrivKey(pub Scalar);
+
+#[derive(Debug, Clone)]
+pub struct Signature {
+    pub s: Scalar,
+    pub R: RistrettoPoint,
+}
+
+#[derive(Clone)]
+pub struct Message(pub Vec<u8>);
+
+impl Signature {
+    pub fn verify(
+        &self,
+        // The message `m` should already have been fed into the transcript
+        transcript: &Transcript,
+        P: VerificationKey,
+    ) -> Result<(), VMError> {
+        let G = RISTRETTO_BASEPOINT_POINT;
+        let mut transcript = transcript.clone();
```

Comment:
i guess this clone is to match the clone inside the prover? Yeah, we definitely need to revisit it in a follow-up PR.

Reply:
yup exactly... if the verifier didn't clone the transcript (if it received a mutable borrow of the transcript), then the transcript states would differ for the prover & verifier at the end of the protocol, which would be bad for future parts.